### PR TITLE
Add recent resource sitemap endpoint for posts

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR adds an endpoint to our sitemap route called `/sitemap-posts.xml`, with an optional number called `/sitemap-posts-2.xml` etc.

Sitemaps need to be at the root of the domain, which is why the logic like this exists in the first place.

Currently we _only_ have "monthly" sitemaps... `/sitemap-Sep-2021.xml`, which forces an admin to go into Google Search Console at the beginning of every month to add a new sitemap.

This allows an admin to provide encapsulated sitemaps and glean information about what Google is searching when, but is not an ideal _default_ sitemap to provide. This is the "default" sitemap for posts on Forem.

There is optionally a zero-indexed value to append here if a Forem has more than 10k published posts, but generally a Forem can get by for a while with just submitting `/sitemap-posts` to Google Search Console.

This naming convention opens us up to adding `/sitemap-pages` and `/sitemap-users`, `/sitemap-organizations`, etc.... But I thought starting here was good, as those would require additional functionality.

Screen Shot 2021-09-29 at 11 13 16 AM

This simplifies SEO and allows us to give clear instructions to creators:

1. Submit `/sitemap-posts` to Google Search Console to help your crawl rate. This will include the most recent 10k indexable posts on the platform.
2. Optionally submit monthly sitemaps if you want to see a breakdown of crawl based on month.

I have chosen to use "posts" as opposed to "articles" which I think of more of a "private term" if we can help it.

## Related Tickets & Documents

There are a couple forem.dev posts about sitemap confusion and this should be a step in helping us provide a good answer.

https://forem.dev/lee/sitemaps-submission-frequency-5ep2
https://forem.dev/akhil/unable-to-submit-sitemap-5g

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [x] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

### CODE REVIEW

File: app/controllers/sitemaps_controller.rb [hunk 1]
```diff
@@ -36,4 +50,8 @@ def set_surrogate_controls(date)
       @stale_while_revalidate = "432000" # five days
     end
   end
+
+  def offset
+    params[:sitemap].split("-")[2].to_i * RESULTS_LIMIT # elvaluates to 0 if not present or not a number
```
Comment (citizen428) [inline app/controllers/sitemaps_controller.rb:55]: There's a typo in the comment, it should read "evaluates". But maybe the code should just be a bit clearer so we don't need a comment at all.

**Review by citizen428 — APPROVED** [review]

File: spec/requests/sitemaps_spec.rb [hunk 2]
```diff
@@ -40,5 +40,39 @@
       expect(response.body).not_to include(articles.last.path)
       expect(response.media_type).to eq("application/xml")
     end
+
+    it "renders most recent posts if /sitemap-posts" do
```
Comment (juliannatetreault) [inline spec/requests/sitemaps_spec.rb:44]: _**Quick, non-blocking suggestion:**_
```suggestion
    it "renders most recent posts if /sitemap-posts", :aggregate_failures do
```

If you decide to move forward with aggregating the failures here, then I think we should also do the same for the other specs that have multiple expectations!

File: spec/requests/sitemaps_spec.rb [hunk 3]
```diff
@@ -40,5 +40,39 @@
...
+      create_list(:article, 8)
+      get "/sitemap-posts.xml"
+      expect(response.body).to include(Article.order("published_at DESC").first.path)
+      expect(response.body).not_to include(Article.order("published_at DESC").last.path)
+    end
+
+    it "renders second page if /sitemap-posts-1" do
```
Comment (juliannatetreault) [inline spec/requests/sitemaps_spec.rb:51]: _**Quick, non-blocking suggestion:**_
```suggestion
    it "renders second page if /sitemap-posts-1", :aggregate_failures do
```

File: spec/requests/sitemaps_spec.rb [hunk 4]
```diff
@@ -40,5 +40,39 @@
...
+      create_list(:article, 8)
+      get "/sitemap-posts-1.xml"
+      expect(response.body).not_to include(Article.order("published_at DESC").first.path)
+      expect(response.body).to include(Article.order("published_at DESC").last.path)
+    end
+
+    it "renders first page if /sitemap-posts-randomn0tnumber" do
```
Comment (juliannatetreault) [inline spec/requests/sitemaps_spec.rb:58]: _**Quick, non-blocking suggestion:**_
```suggestion
    it "renders first page if /sitemap-posts-randomn0tnumber", :aggregate_failures do
```

File: spec/requests/sitemaps_spec.rb [hunk 5]
```diff
@@ -40,5 +40,39 @@
...
+      create_list(:article, 8)
+      get "/sitemap-posts-randomn0tnumber.xml"
+      expect(response.body).to include(Article.order("published_at DESC").first.path)
+      expect(response.body).not_to include(Article.order("published_at DESC").last.path)
+    end
+
+    it "renders empty if /sitemap-posts-2" do
```
Comment (juliannatetreault) [inline spec/requests/sitemaps_spec.rb:65]: _**Quick, non-blocking suggestion:**_
```suggestion
    it "renders empty if /sitemap-posts-2", :aggregate_failures do
```

**Review by juliannatetreault — COMMENTED** [review]
Left a few non-blocking comments, otherwise this is looking good!

**Review by juliannatetreault — APPROVED** [review]

**Pull request merged**